### ISSUE TITLE
Report unrestricted use of runtimeType

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#1475]
Usage of the [`runtimeType`](https://api.dartlang.org/dev/2.2.1-dev.0.0/dart-core/Object/runtimeType.html) property of Object is expensive in dart2js. The compiler has written some optimizations however, so that _restricted_ usage of `runtimeType` are less expensive. The lint rule would report on any usage of `runtimeType` except the following cases:

1. `Type == x.runtimeType`,
2. `a.runtimeType == b.runtimeType` (including `this.runtimeType`), where `a` and `b` are not _both_ `dynamic`,
3. `x.runtimeType.toString()`.

**Comment by srawlins** [comment]
FYI @sigmundch

**Comment by matanlurey** [comment]
re: 2, it should be where either `a` or `b` is `this.runtimeType`, I think.

For example:

```dart
void anExample(Function a, Object b) {
  print(a.runtimeType == b.runtimeType);
}
```

... is going to be a huge de-optimizer despite not violating the rules in 2.

**Comment by sigmundch** [comment]
Agree that it is a good idea to restrict (2) further.

I'm not as familiar with the linter, but if it cannot do checks based on type information, then I would also suggest using `this.runtimeType` (it is easy to do completely syntactically)

If the linter has a way to understand the types of expressions themselves, then we could allow more, but the rules get trickier as @matanlurey showed. We could check whether `a` or `b` is a type that has only non-generic subclasses or that has a relatively small type hierarchy underneath. The main thing is that we want to avoid preserving type-parameters of unrelated types. Defining what is a "relatively small" hierarchy can get confusing as well.

My guess is that the `this.runtimeType` is the simplest and most intuitive way to move forward.

**Comment by srawlins** [comment]
I can first write it to only allow `a.runtimeType == b.runtimeType` if one of `a` or `b` is the identifier `this`. If that is too restrictive, or we can be smarter, we can loosen later. _Tightening_ later gets dodgy though, for projects which break on lint reports.

**Comment by srawlins** [comment]
@davidmorgan suggested a dart2js flag instead. I've filed that as https://github.com/dart-lang/sdk/issues/36154.

Note that we don't have to choose one or the other, but if both the lint and dart2js flag are ideas we like, it would be prudent which to run with first, in order to support customers just clean up existing unrestricted runtimeType usage.
